### ISSUE TITLE
Way to specify additional headers to be sent along ResumableUpload.InitiateSessionAsync()

### ISSUE THREAD

**Issue by Et3rnos** [googleapis/google-cloud-dotnet#12034]
There is currently no way to pass extra headers to be sent alongside ResumableUpload.InitiateSessionAsync().

I'm currently creating a website that should allow users to upload big files directly to the Google Storage bucket, and I would like to specify the size of the file to be uploaded ahead of the time. Your documentation says this is possible by passing a X-Upload-Content-Length header, but after spending some time looking at this package's documentation and even its source, seems like this is not possible.

> From https://cloud.google.com/storage/docs/performing-resumable-uploads#json-api:
> 
> Optional headers that you can add to the request include [X-Upload-Content-Type](https://cloud.google.com/storage/docs/json_api/v1/parameters#xuploadcontenttype) and [X-Upload-Content-Length](https://cloud.google.com/storage/docs/json_api/v1/parameters#xuploadcontentlength).

Are there plans to make this possible? Is there any workaround? Thanks.

**Comment by Et3rnos** [comment]
Looks like a possible workaround is:

```cs
uploader.Service.HttpClient.DefaultRequestHeaders.TryAddWithoutValidation("X-Upload-Content-Length", "1");
```

Would be nice to have a dedicated option for it though.

**Comment by Et3rnos** [comment]
Hi @jskeet and thanks for the quick response. I'll explain why that header is so important for me and give you a bit more of information:

First of all, I'm using Google.Cloud.Storage.V1 (v4.8.0), as you pointed out. This was the .NET client library recommended by the docs page. If you have any other alternative that does what I'm about to explain I'm down for it though :)

I'm building a website that should allow users to upload huge files, but each user also has a quota he has to respect. I'm currently using resumable uploads to upload files, and the flow is similar to the following:

- user manifests his intent to upload a file by sending a request to one of my endpoints, I run something similar to the following on the server:

```cs
var uploader = storageClient.CreateObjectUploader(bucketName, key, request.ContentType, new MemoryStream([]), new UploadObjectOptions
{
      Origin = "<the site url>",
});
var session = await uploader.InitiateSessionAsync();
return Ok(new
{
    Url = session.AbsoluteUri,
});
```

- user gets the session url and uploads the file using a PUT request
- ... (the rest isn't really relevant)

However, as I mentioned, each user has a quota he must respect, if it is for example 100MB, they cannot upload a file bigger than 100MB, at least that's the idea. This is done by setting the `X-Upload-Content-Length` header in the initiation request, which is something you don't currently support.

It is also worth mentioning that the workaround I mentioned in my second comment DOES NOT work, looks like the headers are being overwritten somehow. After the overwrite takes place, the value of `X-Upload-Content-Length` becomes 0, which based on behaviour I'm seeing, means there is no size limit on the file the user can upload through the frontend.

Note that I'm analyzing the requests using Fiddler, so these values are accurate and not assumptions. I'm also replicating the requests in python and I can say for sure that if you could allow arbitrary `X-Upload-Content-Length` header values, what I'm trying to do would be trivial.

Let me know if you need more information.

**Comment by jskeet** [comment]
Thanks for the extra detail. I'll try to come up with a hacky approach for you to test (likely to involve reflection) and if that works, we can work out an appropriate API surface.

**Comment by jskeet** [comment]
Okay, I think I understand a) why your workaround doesn't work; b) how to do this without any changes to the libraries.

Our underlying library (Google.Apis) already supports both of these headers, and populates them when initializing the request, if the length and the content type are known - see https://github.com/googleapis/google-api-dotnet-client/blob/main/Src/Support/Google.Apis/Upload/ResumableUpload.cs#L1170

So if you pass in a stream which claims to have the *right* length, that's what the header value will be set to. Note that this is expected to be *exactly* the right length it's not an upper bound. So your clients will need to tell you the size of the file they want to upload before you initiate the session - you then initiate the session with that length, and Storage will validate that they don't upload more or less than that.

To pass in a stream which claims to have the right length - but no data - you can subclass stream yourself:

```csharp
internal sealed class LengthOnlyStream : Stream
{
    internal LengthOnlyStream(long length) => Length = length;

    public override long Length { get; }
    public override bool CanSeek => true;

    public override bool CanRead => throw new NotImplementedException();
    public override bool CanWrite => throw new NotImplementedException();

    public override long Position
    {
        get => throw new NotImplementedException();
        set => throw new NotImplementedException();
    }

    public override void Flush() => throw new NotImplementedException();
    public override int Read(byte[] buffer, int offset, int count) =>
        throw new NotImplementedException();
    public override long Seek(long offset, SeekOrigin origin) =>
        throw new NotImplementedException();
    public override void SetLength(long value) =>
        throw new NotImplementedException();
    public override void Write(byte[] buffer, int offset, int count) =>
        throw new NotImplementedException();
}
```

I've tested this myself, and it works. It's ugly, but it works. Please could you test this, and if it works for you, we can potentially create a helper method directly on `StorageClient` which initiates an upload with a given length and content type, but without everything else.

**Comment by Et3rnos** [comment]
Hi @jskeet, that definitely solved my problem, thank you! And yes, would be nice to have a helper method for that. Just make it clear for users that if they pass 0 as the length, their upload won't be limited at all, it is doomed to cause some misunderstandings. Otherwise, users attempting to upload 0-byte files might bypass that, it's not something you should worry about but would be nice to have it in the method documentation :P

Well, just a suggestion, thanks for helping!

**Comment by jskeet** [comment]
> Just make it clear for users that if they pass 0 as the length, their upload won't be limited at all, it is doomed to cause some misunderstandings.
> Otherwise, users attempting to upload 0-byte files might bypass that, it's not something you should worry about but would be nice to have it in the method documentation :P

Hmm... I'm tempted to make it nullable, and make a value of 0 fail. (So if you want it to be "any length", just pass in null; if you want a specific length, pass in that length, and we don't support "specifically 0 bytes".)

That sounds like a reasonable thing to do - I can't promise when we'll get to it though.

**Comment by Et3rnos** [comment]
Yes, seems like a good approach to me. The stream workaround works fine so there is no pressure to implement the helper method. I'm not sure if I should close this issue so feel free to do.

**Comment by jskeet** [comment]
Right, this is implemented in 4.9.0 which is in the process of being released now. You'll use `client.InitiateUploadSessionAsync`.
